Write guides to implement new-style functions (#4986)
chainer/chainer: docs/source/guides/functions.rst

.. include:: ../imports.rst

Differentiable Functions
------------------------

Chainer provides a collection of functions in the :mod:`chainer.functions` module.
It covers typical use cases in deep learning, so many existing works can be implemented with them.
On the other hand, deep learning is evolving rapidly and we cannot cover all possible functions to define unseen architectures.
So it is important to learn how to define your own functions.

Old-Style v.s. New-Style Functions
----------------------------------

In Chainer, you can define a function in two ways: old-style and new-style.

* Old-style functions inherit from :class:`chainer.Function` class.
  Forward and backward computation can be implemented using NumPy/CuPy.
* New-style functions inherit from :class:`chainer.FunctionNode` class (introduced in Chainer v3).
  Forward computation can be implemented using NumPy/CuPy.
  Backward computation needs to be implemented using other new-style functions.

> The primary advantage of using new-style functions is that they support computation of higher-order gradient (a.k.a. higher-order derivative or double back-propagation).

Reviewer: gradients

> The primary advantage of using new-style functions is that they support computation of higher-order gradient (a.k.a. higher-order derivative or double back-propagation).

Reviewer: ```suggestion
The primary advantage of using new-style functions is that they support computation of higher-order gradients (a.k.a. higher-order derivative or double back-propagation).
```

> Higher-order gradient is used in some models e.g., recently-proposed GAN architectures.

Reviewer: gradients are used

> Higher-order gradient is used in some models e.g., recently-proposed GAN architectures.

Reviewer: ```suggestion
Higher-order gradients are used in some models e.g., recently-proposed GAN architectures.
```

Currently, most of built-in functions are implemented as :class:`~chainer.FunctionNode` (with a few exceptions listed in `#4449 <https://github.com/chainer/chainer/issues/4449>`__).

> Basically, we suggest you to use new-style when implementing new functions.

Reviewer: We recommend you use

> Basically, we suggest you to use new-style when implementing new functions.

Reviewer: ```suggestion
Basically, we recommend you use new-style when implementing new functions.
```

Here, we just show how to use this flag to make a function support training/test mode.
You will need to check the value of the boolean flag ``chainer.config.train`` and branch appropriately.

For example, consider the following simple dropout function::

  def dropout(x):

>       xp = backend.get_array_module(x.data)

Reviewer 2: ```suggestion
      xp = backend.get_array_module(x.array)
```

      mask = 2 * (xp.random.rand(*x.shape) > 0.5).astype(x.dtype)
      return x * mask

This function applies dropout to each element and doubles survived elements to preserve the scale.
The above implementation applies dropout even in test mode, but it is not a desired behavior.
We can fix it as follows::

  def dropout(x):
      if not chainer.config.train:
          return x

>       xp = backend.get_array_module(x.data)

Reviewer 2: ```suggestion
      xp = backend.get_array_module(x.array)
```
